File: UnityProject/Assets/Scripts/optionsmenuscript.js
#pragma strict

var skin:GUISkin;
var windowRect : Rect;
var menu_width = 300;
var menu_height = 500;
var show_menu = false;

function OnApplicationPause() {  
	Screen.lockCursor = false;
}

function OnApplicationFocus() {
	if(!show_menu){
		Screen.lockCursor = true;
	}
}

function ShowMenu(){
	show_menu = true;
}

function HideMenu(){
	show_menu = false;
}

function OnGUI () {
	if(!show_menu){
		return;
	}
	windowRect = GUI.Window (
		0, 
		Rect(Screen.width*0.5 - menu_width*0.5, Screen.height*0.5 - menu_height*0.5, menu_width, menu_height), 
		WindowFunction, 
		"",
		skin.window);
}

private var draw_cursor : Vector2;
private var draw_cursor_line : Vector2;
private var line_height = 24;
private var line_offset = 24;

function DrawCursor_Init() {
	draw_cursor = Vector2(25,25);
	draw_cursor_line = Vector2(0,0);	
}

function DrawCursor_NextLine() {
	draw_cursor_line = Vector2(0,0);	
	draw_cursor.y += line_offset;
}

function DrawCursor_Offset(val : float) {
	draw_cursor_line.x += val;	
}

function DrawCursor_Get() : Vector2 {
	return draw_cursor + draw_cursor_line;
}

function DrawCursor_RectSpace(width : float) : Rect {
	var rect = Rect(draw_cursor.x + draw_cursor_line.x,
					draw_cursor.y + draw_cursor_line.y,
					width,
					line_height);
	DrawCursor_Offset(width);
	return rect;
}

function DrawLabel(text : String) {
	DrawCursor_Offset(17);
	GUI.Label (
		DrawCursor_RectSpace(400),//GUI.skin.label.CalcSize(new GUIContent(text)).x), 
		text, 
		skin.label);
}

function DrawCheckbox( val : boolean, text : String ) : boolean {
	val = GUI.Toggle (
		DrawCursor_RectSpace(400), 
		val,
		text, 
		skin.toggle);
	return val;			
}

function DrawSlider( val : float) : float {
	DrawCursor_Offset(18);
	val = GUI.HorizontalSlider (
		DrawCursor_RectSpace(400 - draw_cursor_line.x), 
		val,
		0.0, 
		1.0);
	return val;			
}

function DrawButton( text : String) : boolean {
	var val = GUI.Button (
		DrawCursor_RectSpace(200), 
		text,
		skin.button);
	return val;			
}

//private var brightness = 0.3;
private var master_volume = 1.0;
private var sound_volume = 1.0;
private var music_volume = 1.0;
private var voice_volume = 1.0;
private var lock_gun_to_center = false;
private var mouse_invert = false;
private var mouse_sensitivity = 0.2;
private var show_advanced_sound = false;
private var toggle_crouch = true;
private var scroll_view_vector = Vector2(0,0);
private var vert_scroll = 0.0;
private var vert_scroll_pixels = 0.0;
 
function RestoreDefaults() {
	master_volume = 1.0;
	sound_volume = 1.0;
	music_volume = 1.0;
	voice_volume = 1.0;
	mouse_sensitivity = 0.2;
	lock_gun_to_center = false;
	mouse_invert = false;
	toggle_crouch = true;      
}

function Start() {
	Screen.lockCursor = true;
	RestoreDefaults();
	master_volume = PlayerPrefs.GetFloat("master_volume", master_volume);
	sound_volume = PlayerPrefs.GetFloat("sound_volume", sound_volume);
	music_volume = PlayerPrefs.GetFloat("music_volume", music_volume);
	voice_volume = PlayerPrefs.GetFloat("voice_volume", voice_volume);
	mouse_sensitivity = PlayerPrefs.GetFloat("mouse_sensitivity", mouse_sensitivity);
	lock_gun_to_center = PlayerPrefs.GetInt("lock_gun_to_center", lock_gun_to_center?1:0)==1;
	mouse_invert = PlayerPrefs.GetInt("mouse_invert", mouse_invert?1:0)==1;
	toggle_crouch = PlayerPrefs.GetInt("toggle_crouch", toggle_crouch?1:0)==1;      
}
 

function SavePrefs() {
	PlayerPrefs.SetFloat("master_volume", master_volume);
	PlayerPrefs.SetFloat("sound_volume", sound_volume);
	PlayerPrefs.SetFloat("music_volume", music_volume);
	PlayerPrefs.SetFloat("voice_volume", voice_volume);
	PlayerPrefs.SetFloat("mouse_sensitivity", mouse_sensitivity);
	PlayerPrefs.SetInt("lock_gun_to_center", lock_gun_to_center?1:0);
	PlayerPrefs.SetInt("mouse_invert", mouse_invert?1:0);
	PlayerPrefs.SetInt("toggle_crouch", toggle_crouch?1:0);    
}

function IsMenuShown() : boolean {
	return show_menu;
}

function Update() {
	if(vert_scroll != -1.0){
		vert_scroll += Input.GetAxis("Mouse ScrollWheel");
	}
    if(Input.GetKeyDown ("escape") && !show_menu){
        Screen.lockCursor = false;
        ShowMenu();
    } else if(Input.GetKeyDown ("escape") && show_menu){
        Screen.lockCursor = true;
        HideMenu();
    }
    if(Input.GetMouseButtonDown(0) && !show_menu){
        Screen.lockCursor = true;
    }
    if(show_menu){
    	Time.timeScale = 0.0;
    } else {
    	if(Time.timeScale == 0.0){
	    	Time.timeScale = 1.0;
    	}
    }
}

function WindowFunction (windowID : int) {

    scroll_view_vector = GUI.BeginScrollView (
    	Rect (0, 0, windowRect.width, windowRect.height), 
    	scroll_view_vector, 
    	Rect (0, vert_scroll_pixels, windowRect.width, windowRect.height));

	DrawCursor_Init();
	mouse_invert = DrawCheckbox(mouse_invert, "Invert mouse");
	DrawCursor_NextLine();
	DrawLabel("Mouse sensitivity:");
	DrawCursor_NextLine();
	mouse_sensitivity = DrawSlider(mouse_sensitivity);
	DrawCursor_NextLine();
	toggle_crouch = DrawCheckbox(toggle_crouch, "Toggle crouch");
	DrawCursor_NextLine();
	lock_gun_to_center = DrawCheckbox(lock_gun_to_center, "Lock gun to screen center");
	DrawCursor_NextLine();
	/*DrawLabel("Brightness:");
	DrawCursor_NextLine();
	brightness = DrawSlider(brightness);
	DrawCursor_NextLine();*/
	DrawLabel("Master volume:");
	DrawCursor_NextLine();
	master_volume = DrawSlider(master_volume);
	DrawCursor_NextLine();
	show_advanced_sound = DrawCheckbox(show_advanced_sound, "Advanced sound options");
	DrawCursor_NextLine();
	if(show_advanced_sound){
		var indent = 44;
		DrawLabel("....Sounds:");
		DrawCursor_NextLine();
		DrawCursor_Offset(indent);
		sound_volume = DrawSlider(sound_volume);
		DrawCursor_NextLine();
		DrawLabel("....Voice:");
		DrawCursor_NextLine();
		DrawCursor_Offset(indent);
		voice_volume = DrawSlider(voice_volume);
		DrawCursor_NextLine();
		DrawLabel("....Music:");
		DrawCursor_NextLine();
		DrawCursor_Offset(indent);
		music_volume = DrawSlider(music_volume);
		DrawCursor_NextLine();
	}
	if(DrawButton("Resume")){
		Screen.lockCursor = true;
		show_menu = false;
	}
	draw_cursor.y += line_offset * 0.3;
	DrawCursor_NextLine();
	if(DrawButton("Restore defaults")){
		RestoreDefaults();
	}
	DrawCursor_NextLine();	
	draw_cursor.y += line_offset * 0.3;
	if(DrawButton("Exit game")){
		Application.Quit();
	}
	
	var content_height = draw_cursor.y;
	
    GUI.EndScrollView();
    
    if(content_height > windowRect.height){
    	var leeway = (content_height / windowRect.height - windowRect.height / content_height);
    	if(vert_scroll == -1.0){
    		vert_scroll = leeway;
    	}
		vert_scroll = GUI.VerticalScrollbar (
			Rect (menu_width-20, 20, menu_width, menu_height-25), 
			vert_scroll, 
			windowRect.height / content_height,
			content_height / windowRect.height, 
			0.0);
		vert_scroll_pixels = windowRect.height * (leeway - vert_scroll);
	} else {
		vert_scroll = -1.0;
		vert_scroll_pixels = 0.0;
	}
	SavePrefs();
}

File: UnityProject/Assets/Scripts/optionsmenuscript.cs
using UnityEngine;
using UnityEngine.UI;
using UnityEngine.Rendering.PostProcessing;

public class optionsmenuscript : MonoBehaviour {
    public static bool show_menu = false;
    public static bool show_mod_ui = false;

    public GameObject menu;
    public GameObject menuOptions;
    public GameObject optionsContent;
    public Camera uiCamera;
    private GraphicRaycaster raycaster;

    private PostProcessLayer postProcessLayer;
    private PostProcessVolume postProcessVolume;
    private AutoExposure autoExposure;
    private Bloom bloom;
    private Vignette vignette;
    private AmbientOcclusion ambientOcclusion;

    public void OnApplicationPause() {  
        UnlockCursor();
    }

    public void OnApplicationFocus() {
        if(!show_menu) {
            LockCursor();
        }
    }

    private void Awake() {
        raycaster = GetComponent<GraphicRaycaster>();

        postProcessVolume = Camera.main.GetComponent<PostProcessVolume>();
        postProcessLayer = Camera.main.GetComponent<PostProcessLayer>();
        autoExposure = postProcessVolume.profile.GetSetting<AutoExposure>();
        bloom = postProcessVolume.profile.GetSetting<Bloom>();
        vignette = postProcessVolume.profile.GetSetting<Vignette>();
        ambientOcclusion = postProcessVolume.profile.GetSetting<AmbientOcclusion>();
    }

    public void Start() {
        LockCursor();

        if(PlayerPrefs.GetInt("set_defaults", 1) == 1) {
            RestoreDefaults();
        }

        Preferences.UpdatePreferences();
        UpdateUIValuesAndApplyDefaults();
    }

    public void Update() {
        if(Input.GetKeyDown("escape") && !show_menu) {
            ShowMenu();
        } else if(Input.GetKeyDown("escape") && show_menu) {
            HideMenu();
        }

        if(Input.GetMouseButtonDown(0) && !show_menu) {
            LockCursor();
        }

        raycaster.enabled = !ImGuiUnity.instance.MouseGrabbed;
    }

    public void OnGUI() {
        GUI.depth = -20;
        if(show_menu && Event.current.type == EventType.Repaint) {
            uiCamera.Render();
        }
    }

    private void LockCursor() {
        Cursor.lockState = CursorLockMode.Locked;
        Cursor.visible = false;
    }

    private void UnlockCursor() {
        Cursor.lockState = CursorLockMode.None;
        Cursor.visible = true;
    }

    public void ShowMenu() {
        show_menu = true;
        menu.SetActive(true);
        UnlockCursor();
        Time.timeScale = 0.0f;
    }
    
    public void HideMenu() {
        show_menu = false;
        show_mod_ui = false;
        menu.SetActive(false);
        LockCursor();
        Time.timeScale = 1.0f;

        Preferences.UpdatePreferences();
    }

    public static bool IsMenuShown() {
        return show_menu;
    }
    
    public void UpdateInt(Toggle toggle) {
        PlayerPrefs.SetInt(toggle.name, toggle.isOn ? 1 : 0);
    }

    public void UpdateFloat(Slider slider) {
        PlayerPrefs.SetFloat(slider.name, slider.value);
    }

    public void UpdateSwitch(Dropdown dropdown) {
        PlayerPrefs.SetInt(dropdown.name, dropdown.value);
    }

    public void UpdateUIValuesAndApplyDefaults() {
        foreach(Transform transform in optionsContent.transform) {
            if(transform.name.StartsWith("_")) // Don't default settings that start with _
                continue;

            // Update Sliders
            Slider slider = transform.GetComponent<Slider>();
            if(slider != null) {
                if(PlayerPrefs.HasKey(slider.name))
                    slider.SetValueWithoutNotify(PlayerPrefs.GetFloat(slider.name));
                slider.onValueChanged.Invoke(slider.value);
                continue; // Don't need to check for other Setting types
            }

            // Update Toggles
            Toggle toggle = transform.GetComponent<Toggle>();
            if(toggle != null) {
                if(PlayerPrefs.HasKey(toggle.name))
                    toggle.SetIsOnWithoutNotify(PlayerPrefs.GetInt(toggle.name) == 1);
                toggle.onValueChanged.Invoke(toggle.isOn);
                continue;
            }

            // Update Dropdowns
            Dropdown dropdown = transform.GetComponent<Dropdown>();
            if(dropdown != null) {
                if(PlayerPrefs.HasKey(dropdown.name))
                    dropdown.SetValueWithoutNotify(PlayerPrefs.GetInt(dropdown.name));
                dropdown.onValueChanged.Invoke(dropdown.value);
                continue;
            }
        }
    }

    public void RestoreDefaults() {
        PlayerPrefs.SetInt("set_defaults", 0);

        PlayerPrefs.SetFloat("master_volume", 1.0f);
        PlayerPrefs.SetFloat("sound_volume", 1.0f);
        PlayerPrefs.SetFloat("music_volume", 1.0f);
        PlayerPrefs.SetFloat("voice_volume", 1.0f);
        PlayerPrefs.SetFloat("mouse_sensitivity", 0.2f);
        PlayerPrefs.SetFloat("gun_distance", 1f);
        PlayerPrefs.SetInt("lock_gun_to_center", 0);
        PlayerPrefs.SetInt("mouse_invert", 0);
        PlayerPrefs.SetInt("toggle_crouch", 1);

        PlayerPrefs.SetFloat("post_processing", 1f);
        PlayerPrefs.SetFloat("ambient_intensity", 0.44f);
        PlayerPrefs.SetFloat("bloom_intensity", 1f);

        PlayerPrefs.SetFloat("auto_exposure_min_luminance", -3.3f);
        PlayerPrefs.SetFloat("auto_exposure_max_luminance", 2.8f);
        PlayerPrefs.SetFloat("auto_exposure_exposure_compensation", 0.93f);
        
        PlayerPrefs.SetInt("antialiasing_mode", 3);
        PlayerPrefs.SetInt("vignette", 0);

        PlayerPrefs.SetInt("ignore_vanilla_guns", 0);
        PlayerPrefs.SetInt("ignore_vanilla_tiles", 0);
    }

    // Functionality
    public void ToggleOptions() {
        menuOptions.SetActive(!menuOptions.activeSelf);
    }

    public void ExitGame() {
        UnityEngine.Application.Quit();
    }

    public void SetPostProcessingEnabled(Toggle toggle) {
        postProcessLayer.enabled = toggle.isOn;
    }
    
    public void SetPostProcessingWeight(float weight) {
        postProcessVolume.weight = weight;
    }

    public void SetAmbientIntensity(float intensity) {
        ambientOcclusion.intensity.Override(intensity);
    }

    public void SetBloomIntensity(float intensity) {
        bloom.intensity.Override(intensity);
    }

    public void SetAutoExposureMinEV(float autoExposureMinLuminance) {
        autoExposure.minLuminance.Override(autoExposureMinLuminance);
    }

    public void SetAutoExposureMaxEV(float autoExposureMaxLuminance) {
        autoExposure.maxLuminance.Override(autoExposureMaxLuminance);
    }

    public void SetAutoExposureExposureCompensation(float autoExposureExposureCompensation) {
        autoExposure.keyValue.Override(autoExposureExposureCompensation);
    }
 
    public void SetAAMode(int mode) {
        postProcessLayer.antialiasingMode = (PostProcessLayer.Antialiasing) mode;
    }

    public void SetVignette(bool enabled) {
        vignette.active = enabled;
    }

    public void SetVSync(bool enabled) {
        QualitySettings.vSyncCount = enabled ? 1 : 0;
    }

    public void OpenModsFolder() {
        Application.OpenURL($"\"{ModManager.GetModsfolderPath()}\"");
    }

    public void ToggleModUI() {
        show_mod_ui = !show_mod_ui;
    }

    public void ReopenCurrentScene() {
        UnityEngine.SceneManagement.SceneManager.LoadScene(UnityEngine.SceneManagement.SceneManager.GetActiveScene().name);
        HideMenu();
    }

    public void OpenScene(string scene_name) {
        UnityEngine.SceneManagement.SceneManager.LoadScene(scene_name);
        HideMenu();
    }
}
